packages/@sanity/schema/src/sanity/extractSchema.ts

```diff
@@ -0,0 +1,530 @@
+  stringSchemaType: StringSchemaType,
+): StringTypeNode | UnionTypeNode<StringTypeNode> {
+  if (stringSchemaType.options?.list) {
+    return {
+      type: 'union',
+      of: stringSchemaType.options.list.map((v) => ({
+        type: 'string',
+        value: typeof v === 'string' ? v : v.value,
+      })),
+    }
+  }
+  return {
+    type: 'string',
+  }
+}
+
+function createNumberTypeNodeDefintion(
+  numberSchemaType: NumberSchemaType,
+): NumberTypeNode | UnionTypeNode<NumberTypeNode> {
+  if (numberSchemaType.options?.list) {
+    return {
+      type: 'union',
+      of: numberSchemaType.options.list.map((v) => ({
+        type: 'number',
+        value: typeof v === 'number' ? v : v.value,
+      })),
+    }
+  }
+  return {
+    type: 'number',
+  }
+}
+
+function createReferenceTypeNodeDefintion(
+  reference: ReferenceSchemaType,
+): ObjectTypeNode | UnionTypeNode<ObjectTypeNode> {
+  const references = gatherReferenceNames(reference)
+  if (references.length === 1) {
+    return createReferenceTypeNode(references[0])
+  }
+
+  return {
+    type: 'union',
+    of: references.map((name) => createReferenceTypeNode(name)),
+  }
+}
+
+// Traverse the reference type tree and gather all the reference names
+function gatherReferenceNames(type: ReferenceSchemaType): string[] {
+  const allReferences = gatherReferenceTypes(type)
+  // Remove duplicates
+  return [...new Set([...allReferences.map((ref) => ref.name)])]
+}
+
+function gatherReferenceTypes(type: ReferenceSchemaType): ObjectSchemaType[] {
+  const refTo = 'to' in type ? type.to : []
+  if ('type' in type && isReferenceType(type.type)) {
+    return [...gatherReferenceTypes(type.type), ...refTo]
+  }
+
+  return refTo
+}
+
+// Traverse the type tree and gather all the fields
+function gatherFields(type: SanitySchemaType | ObjectSchemaType): ObjectField[] {
+  if ('fields' in type) {
+    return type.type ? gatherFields(type.type).concat(type.fields) : type.fields
+  }
+
+  return []
+}
+
+// Traverse the type tree and check if the type or any of its subtypes are of the given type
+function isType(
+  typeDef: SanitySchemaType | ObjectField | ObjectFieldType,
+  typeName: string,
+): boolean {
+  let type: SchemaType | ObjectField | ObjectFieldType | undefined = typeDef
+  while (type) {
+    if (type.name === typeName || (type.type && type.type.name === typeName)) {
+      return true
+    }
+
+    type = type.type
+  }
+  return false
+}
+
+// Traverse the type tree and return the "last" type, ie deepest type in the tree
+function lastType(typeDef: SanitySchemaType): SanitySchemaType | undefined {
+  let type: SchemaType | ObjectField | ObjectFieldType | undefined = typeDef
+  while (type) {
+    if (!type.type) {
+      return type
+    }
+    type = type.type
+  }
+
+  return undefined
+}
+
+// Sorts the types by their dependencies by using a topological sort depth-first algorithm.
+function sortByDependencies(compiledSchema: SchemaDef): string[] {
```

Comment:
Very nice – I can easily see this being useful as a util on its own (but let's keep it local here for now).

Reply:
I don't see a use for it right now, so think its fine to leave it as-is. I think the comment and name already includes the right keywords for someone to be able search for it if they're wondering if we got a util for it already.

Reply:
Cool!